src/gsym/parser.rs

```diff
@@ -0,0 +1,339 @@
+//! * String Table
+//! * Address Data
+//!
+//! The standalone GSYM starts with a Header, which describes the
+//! size of an entry in the address table, the number of entries in
+//! the address table, and the location and the size of the string
+//! table.
+//!
+//! Since the Address Table is immediately after the Header, the
+//! Header describes only the size of an entry and number of entries
+//! in the table but not where it is.  The Address Table comprises
+//! addresses of symbols in the ascending order, so we can find the
+//! symbol an address belonging to by doing a binary search to find
+//! the most close address but smaller or equal.
+//!
+//! The Address Data Offset Table has the same number of entries as
+//! the Address Table.  Every entry in one table will has
+//! corresponding entry at the same offset in the other table.  The
+//! entries in the Address Data Offset Table are always 32bits
+//! (4bytes.)  It is the file offset to the respective Address
+//! Data. (AddressInfo actually)
+//!
+//! An AddressInfo comprises the size and name of a symbol.  The name
+//! is an offset in the string table.  You will find a null terminated
+//! C string at the give offset.  The size is the number of bytes of
+//! the respective object; ex, a function or variable.
+//!
+//! See https://reviews.llvm.org/D53379
+//!
+//! # Examples
+//!
+//! The following is an example to find the symbol of an address.
+//!
+//! ```ignore
+//! let ctx = parse_header(gsym_file_content).unwrap();
+//! for i in (0..ctx.num_addresses()) {
+//!     let addr = ctx.addr_at(i);
+//!     println!("{:x}", addr);
+//! }
+//!
+//! // Find the symbol for the address 0xdeadbeef
+//! let addr_idx = find_address(&ctx, 0xdeadbeef);
+//! let start_addr = ctx.addr_at(addr_idx);
+//! if addr < start_add) {
+//!     eprintln!("Not found");
+//! } else {
+//!     let info = ctx.addr_info(addr_idx);
+//!     if addr >= (start_addr + info.size as u64) {
+//!         eprintln!("Not found");
+//!     } else {
+//!         let name = ctx.get_str(info.name);
+//!         println!("symbol name of 0xdeadbeef: {}", name);
+//!     }
+//! }
+//! ```
+use super::types::*;
+
+use std::io::{Error, ErrorKind};
+
+use crate::tools::{decode_udword, decode_uhalf, decode_uword};
+use std::ffi::CStr;
+
+/// Hold the major parts of a standalone GSYM file.
+///
+/// GsymContext provides functions to access major entities in GSYM.
+/// GsymContext can find respective AddressInfo for an address.  But,
+/// it doesn't parse AddressData to get line numbers.
+pub struct GsymContext<'a> {
+    header: Header,
+    addr_tab: &'a [u8],
+    addr_data_off_tab: &'a [u8],
+    str_tab: &'a [u8],
+    raw_data: &'a [u8],
+}
+
+/// Parse the Header of a standalone GSYM file.
+///
+/// # Arguments
+///
+/// * `data` - is the content of a standalone GSYM.
+///
+/// Returns a GsymContext, which includes the Header and other important tables.
+pub fn parse_header(data: &[u8]) -> Result<GsymContext, Error> {
```

Comment:
Also, it seems you are omitting the lifetime on `GsymContext`. I don't believe that syntax is allowed in recent editions of Rust anymore. Would be good to check this in CI.

Reply:
I build it with v1.65.0, and it still works.  Do you have any source about the plan of retiring this?

Reply:
Just searched...it's the `elided_lifetimes_in_paths` lint. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=47a51a41c722fce7f0999c759e7deac0

I was under the impression it was default by now, but one may have to opt in.

Reply:
Ok! Even so, I will add a lifetime in case.

Reply:
Sounds good. I think it's probably captures by a meta lint such as `future_incompatible` (or `rust_2018_compatibility` or whatever; haven't checked). May make sense to enable that at some point.